controller/tests/controller_server/csi_controller_server_test.py

```diff
@@ -118,21 +123,42 @@ def test_create_snapshot_with_empty_name(self, a_enter):
 
     def _prepare_create_snapshot_mocks(self, storage_agent):
         storage_agent.return_value = self.storage_agent
-
+        self.mediator.get_snapshot = Mock()
+        self.mediator.get_snapshot.return_value = None
         self.mediator.create_snapshot = Mock()
         self.mediator.create_snapshot.return_value = utils.get_mock_mediator_response_snapshot(10, snapshot_name, "wwn",
                                                                                                snapshot_volume_name,
                                                                                                "xiv")
 
-    @patch("controller.controller_server.csi_controller_server.get_agent")
-    def test_create_snapshot_succeeds(self, storage_agent):
+    def _test_create_snapshot_succeeds(self, storage_agent, expected_pool=None):
         self._prepare_create_snapshot_mocks(storage_agent)
 
         self.servicer.CreateSnapshot(self.request, self.context)
 
         self.assertEqual(self.context.code, grpc.StatusCode.OK)
-        self.mediator.get_snapshot.assert_called_once_with(snapshot_volume_wwn, snapshot_name, pool=None)
-        self.mediator.create_snapshot.assert_called_once_with(snapshot_volume_wwn, snapshot_name, None)
+        self.mediator.get_snapshot.assert_called_once_with(snapshot_volume_wwn, snapshot_name, pool=expected_pool)
+        self.mediator.create_snapshot.assert_called_once_with(snapshot_volume_wwn, snapshot_name, expected_pool)
+
+    @patch("controller.controller_server.csi_controller_server.get_agent")
+    def test_create_snapshot_succeeds(self, storage_agent):
+        self._test_create_snapshot_succeeds(storage_agent)
+
+    @patch("controller.controller_server.csi_controller_server.get_agent")
+    def test_create_snapshot_with_pool_parameter_succeeds(self, storage_agent):
+        self.request.parameters = {config.PARAMETERS_POOL: pool}
+        self._test_create_snapshot_succeeds(storage_agent, expected_pool=pool)
+
+    def _test_create_snapshot_with_pools_by_system_parameter(self, storage_agent, system_id, expected_pool):
+        self.request.source_volume_id = "{}:{}:{}".format("A9000", system_id, snapshot_volume_wwn)

```

Comment:
I thought the max system type length is 4

Reply:
right, done (thought of XIV)

Reply:
then perhaps this also should be computed (e.g. `max(map(len, ALL_ARRAY_TYPES)))`)

Reply:
where?

Reply:
[here](https://github.com/IBM/ibm-block-csi-driver/pull/324/files#diff-805936c9666f06d2a0664c47d089a0d84d457d1f07cda4d80713d8e5bfb3eefaR15)

Reply:
ok, done

Reply:
I think it would be easier if we maintain a `ALL_ARRAY_TYPES` const in the same file the types are defined

Reply:
done